zmq4_routerdealer_test.go

```diff
@@ -255,34 +255,48 @@ func TestRouterWithNoDealer(t *testing.T) {
 	}
 }
 
-func TestRouterClose(t *testing.T) {
-	ctx, cancel := context.WithCancel(context.Background())
-	defer cancel()
-	router := zmq4.NewRouter(ctx)
-	err := router.Listen("tcp://*:*")
-	if err != nil {
-		t.Fatalf("router could not listen: %+v", err)
-	}
-	_, port, _ := net.SplitHostPort(router.Addr().String())
-	dealer := zmq4.NewDealer(ctx)
-	err = dealer.Dial("tcp://*:" + port)
-	if err != nil {
-		t.Fatalf("dealer could not dial: %+v", err)
+func TestRouterDealerClose(t *testing.T) {
+	tests := []struct {
+		name string
+	}{
+		{name: "router"},
+		{name: "dealer"},
 	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			ctx, cancel := context.WithCancel(context.Background())
+			defer cancel()
+			socks := map[string]zmq4.Socket{
+				"router": zmq4.NewRouter(ctx),
+				"dealer": zmq4.NewDealer(ctx),
+			}
+			router := socks["router"]
+			dealer := socks["dealer"]
 
-	var wg sync.WaitGroup
-	wg.Add(1)
-	go func() {
-		defer wg.Done()
-		_, err := router.Recv()
-		if err == nil {
-			t.Errorf("expected error: context canceled")
-		}
-	}()
+			err := router.Listen("tcp://*:*")
+			if err != nil {
+				t.Fatalf("router could not listen: %+v", err)
+			}
+			_, port, _ := net.SplitHostPort(router.Addr().String())
+			err = dealer.Dial("tcp://*:" + port)
+			if err != nil {
+				t.Fatalf("dealer could not dial: %+v", err)
+			}
+			var wg sync.WaitGroup
+			wg.Add(1)
+			go func(sock zmq4.Socket) {
```

Comment:
ITSM there's a race here: depending on how fast/slow that goroutine is scheduled, `sock.Recv` might be issued before the socket is closed.
perhaps we could add another barrier before `sock.Recv`, enabled after `socks[tt.name].Close` has been called?